doc/man/gnunet-download.1
.\" This file is part of GNUnet.
.\" Copyright (C) 2001-2019 GNUnet e.V.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the file
.\" FDL-1.3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/fdl.html}.
.\"
.\" Alternately, this document is also available under the General
.\" Public License, version 3 or later, as published by the Free Software
.\" Foundation.  A copy of the license is included in the file
.\" GPL3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/gpl.html
.\"
.\" SPDX-License-Identifier: GPL3.0-or-later OR FDL1.3-or-later
.\"
.Dd January 31, 2016
.Dt GNUNET-DOWNLOAD 1
.Os
.Sh NAME
.Nm gnunet-download
.Nd a command line interface for downloading files from GNUnet
.Sh SYNOPSIS
.Nm
.Op Fl a Ar LEVEL | Fl -anonymity= Ns Ar LEVEL
.Op Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
.Op Fl D | -delete-incomplete
.Op Fl h | -help
.Op Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
.Op Fl n | -no-network
.Op Fl o Ar FILENAME | Fl -output= Ns Ar FILENAME
.Op Fl p Ar DOWNLOADS | Fl -parallelism= Ns Ar DOWNLOADS
.Op Fl r Ar REQUESTS | Fl -request-parallelism= Ns Ar REQUESTS
.Op Fl R | -recursive
.Op Fl v | -version
.Op Fl V | -verbose
.Ao Ar GNUNET_URI Ac
.Sh DESCRIPTION
Download files from GNUnet.
The options are as follows:
.Bl -tag -width indent
.It Fl a Ar LEVEL | Fl -anonymity= Ns Ar LEVEL
This option can be used to specify additional anonymity constraints.
The default is 1.
If set to 0, GNUnet will publish the file non-anonymously and in fact sign the advertisement for the file using your peer's private key.
This will allow other users to download the file as fast as possible, including using non-anonymous methods (discovery via DHT and CADET transfer).
If you set it to 1 (default), you use the standard anonymous routing algorithm (which does not explicitly leak your identity).
However, a powerful adversary may still be able to perform traffic analysis (statistics) to over time discovery your identity.
You can gain better privacy by specifying a higher level of anonymity (using values above 1).
This tells FS that it must hide your own requests in equivalent-looking cover traffic.
This should confound an adversaries traffic analysis, increasing the time and effort it would take to discover your identity.
However, it also can significantly reduce performance, as your requests will be delayed until sufficient cover traffic is available.
The specific numeric value (for anonymity levels above 1) is simple:
Given an anonymity level L (above 1), each request FS makes on your behalf must be hidden in L-1 equivalent requests of cover traffic (traffic your peer routes for others) in the same time-period.
The time-period is twice the average delay by which GNUnet artificially delays traffic.
Note that regardless of the anonymity level you choose, peers that cache content in the network always use anonymity level 1.
.It Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
Use config file (default:
.Pa ~/.config/gnunet.conf Ns )
.It Fl D | -delete-incomplete
Causes gnunet-download to delete incomplete downloads when aborted with CTRL-C.
Note that complete files that are part of an incomplete recursive download will not be deleted even with this option.
Without this option, terminating gnunet-download with a signal will cause incomplete downloads to stay on disk.
If gnunet-download runs to (normal) completion finishing the download, this option has no effect.
.It Fl h | -help
Print the help page.
.It Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
Change the loglevel.
Possible values for LOGLEVEL are ERROR, WARNING, INFO and DEBUG.
.It Fl n | -no-network
Only search locally, do not forward requests to other peers.
.It Fl o Ar FILENAME | Fl -output= Ns Ar FILENAME
Write the file to
.Ar FILENAME .
Hint: when recursively downloading a directory, append a '/' to the end of the FILENAME to create a directory of that name.
If no FILENAME is specified, gnunet-download constructs a temporary ID from the URI of the file.
The final filename is constructed based on meta-data extracted using
.Xr libextractor 1
(if available).
.It Fl p Ar DOWNLOADS | Fl -parallelism= Ns Ar DOWNLOADS
Set the maximum number of allowed parallel downloads to
.Ar DOWNLOADS .
More parallel downloads can, to some extent, improve the overall time to download content.
However, parallel downloads also take more memory (see also
.Fl r
which can be used to limit memory utilization) and more sockets.
This option is used to limit the number of files that are downloaded in parallel.
.Fl r
can be used to limit the number of blocks that are concurrently requested.
As a result, the value only matters for recursive downloads.
The default value is 32.
.It Fl r Ar REQUESTS | Fl -request-parallelism= Ns Ar REQUESTS
Set the maximum number of parallel requests that is allowed.
If multiple files are downloaded, gnunet-download will not run them in parallel if this would cause the number of pending requests to possibly exceed the given value.
This is useful since, for example, downloading dozens of multi-gigabyte files in parallel could exhaust memory resources and would hardly improve performance.
Note that the limit only applies to this specific process and that other download activities by other processes are not included in this limit.
Consider raising this limit for large recursive downloads with many large files if memory and network bandwidth are not fully utilized and if the parallelism limit
.Pq Fl p
is not reached.
This option also only matters for recursive downloads.
The default value is 4092.
.It Fl R | -recursive
Download directories recursively (and in parallel).
Note that the URI must belong to a GNUnet directory and that the filename given to
.Fl o
must end in '.gnd' \(em otherwise, you will receive an error.
You may want to use "DIRNAME/.gnd" for the filename, this way a directory "DIRNAME/" will be created, and GNUnet's internal directory information will be stored in "DIRNAME/.gnd".
However, it is also possible to specify "DIRNAME.gnd", in which case the files from the directory will end up in "DIRNAME/", while GNUnet's directory meta data will be in "DIRNAME.gnd".
.It Fl v | -version
Print the version number.
.It Fl V | -verbose
Print progress information.
.El
.Ss NOTES
The GNUNET_URI is typically obtained from
.Xr gnunet-search 1 .
.Xr gnunet-fs-gtk 1
can also be used instead of gnunet-download.
If you ever have to abort a download, you can at any time continue it by re-issuing gnunet-download with the same filename.
In that case GNUnet will not download blocks again that are already present.
GNUnet's file-encoding will ensure file integrity, even if the existing file was not downloaded from GNUnet in the first place.
Temporary information will be appended to the target file until the download is completed.
.Sh FILES
.Pa ~/.config/gnunet.conf
GNUnet configuration file
.Sh SEE ALSO
.Xr gnunet-fs-gtk 1 ,
.Xr gnunet-publish 1 ,
.Xr gnunet-search 1 ,
.Xr gnunet-service-fs 1 ,
.Xr gnunet.conf 5
.sp
The full documentation for gnunet is maintained as a Texinfo manual.
If the
.Xr info 1
and gnunet programs are properly installed at your site, the command
.Pp
.Dl info gnunet
.Pp
should give you access to the complete handbook,
.Pp
.Dl info gnunet-c-tutorial
.Pp
will give you access to a tutorial for developers.
.sp
Depending on your installation, this information is also available in
.Xr gnunet 7 and
.Xr gnunet-c-tutorial 7 .
.\".Sh HISTORY
.\".Sh AUTHORS
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt gnunet-developers@gnu.org .


ChangeLog
Wed Oct 30 10:50:00 UTC 2019
  Released gnunet 0.11.8
  -ng0

Sun Oct 27 17:17:00 2019 +0200
  Released gnunet 0.11.7
  -schanzen, ng0

Fri Oct 25 00:00:00 UTC 2019
  Dropped glpk dependency(-check). -ng0

Wed Oct 23 20:55:12 2019 +0200
  Make REST service start per user in a multiuser setup.
  -schanzen

Wed Oct 23 00:00:00 UTC 2019
  Remove setuid helpers. They never worked as intended. The
  fixed version is not portable and the defacto good way to
  handle this which is most portable is to let package managers
  handle this (as they do right now, the good ones), and let
  people handle this who read the documentation.
  This commit removes what would be patched out by the majority
  of package managers, which is setuid handling in Makefiles.
  It is very likely that no one will notice this code is gone.
  -ng0

Thu Oct 17 00:00:00 UTC 2019
  Added support for doas, use it in some places conditionally
  if sudo is not present. Fixed sudo detection. -ng0

Mon Oct 14 00:00:00 UTC 2019
  Added 'pretty' make rule to run uncrustify over the source tree.
  -ng0

Mon Sep 16 00:00:00 UTC 2019
  Remove checks for CXX and CPP. Scans found no
  (remaining) cxx code. ccache with clang on
  NetBSD 9 hit no rebuilds. -ng0

Mon Sep 10 00:00:00 UTC 2019
  Drop win32 and mingw support after discussion with
  Christian. It has been unmaintained in our code base
  for years and there are no known users.
  Future implementors for Windows 10 support can
  refer to this commit as reference. -ng0

Mon Sep 9 00:00:00 UTC 2019
  Remove plibc, closes #5877 as discussed in
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00002.html
  -ng0

Sun Sep 1 13:25:07 2019 +0200
  Improved support for external applications/services wrt
  localization/gettext.
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00000.html

Tue 20 Aug 2019 04:32:09 PM CEST
  Fixed fatal bug in gnunet-bcd that prevented the main page
  from showing up. -CG

Mon Aug 19 17:36:48 2019 +0200
  REST: Added API plugin to read/write GNUnet configuration options -alexia

Sun Jul 28 00:00:00 2019 +0000
  Adding --enable-texinfo4 configure switch, defaulting to off.
  This allows distributors to override the texinfo 5.x and later
  commandline options we use, and allows us on the serverside to
  build html with newer output options. -ng0

Wed 24 Jul 2019 05:43:43 PM CEST
  Releasing GNUnet 0.11.6. -CG

Mon Jul 15 09:57:20 2019 +0200
  The REST server now has a configurable CORS Origin setting -schanzen

Fri Jul 5 22:11:04 2019 +0200
  Added detection of SUID binaries
  through GNUNET_OS_get_suid_binary_path -lurchi

Sat 29 Jun 2019 10:18:55 AM CEST
  Renamed "heap" namestore backend to "flat" as it does
  persist data in a file. Made it work with databases larger
  than 16 MB. -CG

Fri Jun 28 14:10:53 2019 +0200
  RECLAIM: Reissue of tickets (same attributes, same identity)
  will no longer create a new ticket, but reuse the old. -alexia

Thu Jun 27 10:34:46 2019 +0200
  The REST server now returns the Origin header value in the
  CORS header reponse. -schanzen

Thu Jun 27 09:08:06 2019 +0200
  GNS NSS plugin will not timeout and not hang forever and block
  gethostbyname. Final resolution pending #5782 -schanzen

Sat Jun 22 10:38:43 2019 +0200
  Add gnutls support for GNS CA certificate generation -schanzen

Sat 15 Jun 2019 11:45:35 PM CEST
  Improved gnunet-gns performance for many ego scenario. -CG

Fri 14 Jun 2019 07:17:40 PM CEST
  Add option to gnunet-identity to display private keys. -CG

Wed Jun 5 21:19:46 2019 +0200
 Releasing GNUnet 0.11.5. -schanzen

Wed May 29 11:17:49 2019 +0200
  Added CAA record type (RFC 6844). -schanzen

Sat 25 May 2019 08:27:39 AM CEST
  Significantly speed-up ego creation and deletion via
  gnunet-identity command when there are many (hundreds)
  of egos at the same peer. -CG

Sun May 12 12:00:00 2019 +0200
  Releasing GNUnet 0.11.4. -MS

Thu May 9 23:35:24 2019 +0200
  gnunet-identity can now output the public key on creation of new
  egos, and -e can act as a filter when listing all egos. -CG

Thu May 2 08:44:37 2019 +0200
  gnunet-arm now logs to file by default -CG

Sat Apr 27 21:50:57 2019 +0200
  RECLAIM: Added ECDH encryption for authorization code payload -schanzen

Sat 27 Apr 2019 06:36:31 AM CEST
  Bump libmicrohttpd version requirement as some intermediate
  versions have relevant bugs that cause problems. -CG

Sat Apr 27 01:50:43 2019 +0200
  RECLAIM: No longer experimental -schanzen

Fri Apr 26 11:08:21 2019 +0200
  REST: Configuration plugin added -schanzen

Tue Apr 23 23:00:00 UTC 2019
  All manpages are properly licensed. -ng0

Sun Apr 21 22:22:22 UTC 2019
  All manpages are now in mdoc format. -ng0

Sat Apr 20 18:38:43 UTC 2019
  Remove optional gnunet-download-manager.scm and
  with it the optional dependency on a no longer
  maintained version of GNU Guile (#5682). -ng0

Sun Apr 14 09:54:20 2019 +0200
  RECLAIM: Moved away from ABE to GNS encryption. Minor API changed and
  renaming of records. Deprecated SQLite plugin in favor of namestore.
  -schanzen

Sun Apr 7 14:40:59 2019 +0200
  Releasing GNUnet 0.11.3. -CG

Sat 06 Apr 2019 02:36:17 PM CEST
  Introducing GNUNET_memcmp() and GNUNET_is_zero() for better typing
  when comparing memory areas. -CG

Fri 05 Apr 2019 11:38:53 PM CEST
  Use paths of sysctl, ip, ifconfig and ip(6)tables from configure
  if they work. (#5352). -CG

Thu 04 Apr 2019 12:39:44 PM CEST
  Hotfix for build issue. Releasing GNUnet 0.11.2. -CG/MS

Wed 03 Apr 2019 03:18:43 PM CEST
  Releasing GNUnet 0.11.1. -CG

Wed 03 Apr 2019 01:54:44 PM CEST
  Merged gnunet-qr-c-2 branch, replacing Python-based implementation
  of gnunet-qr with C-based implementation that also integrates
  gnunet-uri. -HG/MS/CG

Wed 03 Apr 2019 01:15:57 PM CEST
  Fixed a few test case regressions (GNS, FS). -CG

Mon Apr 1 22:00:00 UTC 2019
  documentation/buildsystem: manpages and other documentation parts
  can now be build more controlled again, this fixes the regression

src/fs/gnunet-download.c
/*
     This file is part of GNUnet.
     Copyright (C) 2001, 2002, 2004, 2005, 2006, 2007, 2009 GNUnet e.V.

     GNUnet is free software: you can redistribute it and/or modify it
     under the terms of the GNU Affero General Public License as published
     by the Free Software Foundation, either version 3 of the License,
     or (at your option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Affero General Public License for more details.

     You should have received a copy of the GNU Affero General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>.

     SPDX-License-Identifier: AGPL3.0-or-later
 */
/**
 * @file fs/gnunet-download.c
 * @brief downloading for files on GNUnet
 * @author Christian Grothoff
 * @author Krista Bennett
 * @author James Blackwell
 * @author Igor Wronsky
 */
#include "platform.h"
#include "gnunet_fs_service.h"

static int ret;

static unsigned int verbose;

static int delete_incomplete;

static const struct GNUNET_CONFIGURATION_Handle *cfg;

static struct GNUNET_FS_Handle *ctx;

static struct GNUNET_FS_DownloadContext *dc;

static unsigned int anonymity = 1;

static unsigned int parallelism = 16;

static unsigned int request_parallelism = 4092;

static int do_recursive;

static char *filename;

static int local_only;


static void
cleanup_task (void *cls)
{
  GNUNET_FS_stop (ctx);
  ctx = NULL;
}


static void
shutdown_task (void *cls)
{
  if (NULL != dc)
  {
    GNUNET_FS_download_stop (dc, delete_incomplete);
    dc = NULL;
  }
}


/**
 * Display progress bar (if tty).
 *
 * @param x current position in the download
 * @param n total size of the download
 * @param w desired number of steps in the progress bar
 */
static void
display_bar (unsigned long long x, unsigned long long n, unsigned int w)
{
  char buf[w + 20];
  unsigned int p;
  unsigned int endeq;
  float ratio_complete;

  if (0 == isatty (1))
    return;
  ratio_complete = x / (float) n;
  endeq = ratio_complete * w;
  GNUNET_snprintf (buf, sizeof(buf), "%3d%% [", (int) (ratio_complete * 100));
  for (p = 0; p < endeq; p++)
    strcat (buf, "=");
  for (p = endeq; p < w; p++)
    strcat (buf, " ");
  strcat (buf, "]\r");
  printf ("%s", buf);
  fflush (stdout);
}


/**
 * Called by FS client to give information about the progress of an
 * operation.
 *
 * @param cls closure
 * @param info details about the event, specifying the event type
 *        and various bits about the event
 * @return client-context (for the next progress call
 *         for this operation; should be set to NULL for
 *         SUSPEND and STOPPED events).  The value returned
 *         will be passed to future callbacks in the respective
 *         field in the `struct GNUNET_FS_ProgressInfo`
 */
static void *
progress_cb (void *cls, const struct GNUNET_FS_ProgressInfo *info)
{
  char *s;
  const char *s2;
  char *t;

  switch (info->status)
  {
  case GNUNET_FS_STATUS_DOWNLOAD_START:
    if (verbose > 1)
      fprintf (stderr,
               _ ("Starting download `%s'.\n"),
               info->value.download.filename);
    break;

  case GNUNET_FS_STATUS_DOWNLOAD_PROGRESS:
    if (verbose)
    {
      s = GNUNET_strdup (
        GNUNET_STRINGS_relative_time_to_string (info->value.download.eta,
                                                GNUNET_YES));
      if (info->value.download.specifics.progress.block_download_duration
          .rel_value_us == GNUNET_TIME_UNIT_FOREVER_REL.rel_value_us)
        s2 = _ ("<unknown time>");
      else
        s2 = GNUNET_STRINGS_relative_time_to_string (info->value.download
                                                     .specifics.progress
                                                     .block_download_duration,
                                                     GNUNET_YES);
      t = GNUNET_STRINGS_byte_size_fancy (
        info->value.download.completed * 1000LL
        / (info->value.download.duration.rel_value_us + 1));
      fprintf (
        stdout,
        _ (
          "Downloading `%s' at %llu/%llu (%s remaining, %s/s). Block took %s to download\n"),
        info->value.download.filename,
        (unsigned long long) info->value.download.completed,
        (unsigned long long) info->value.download.size,
        s,
        t,
        s2);
      GNUNET_free (s);
      GNUNET_free (t);
    }
    else
    {
      display_bar (info->value.download.completed,
                   info->value.download.size,
                   60);
    }
    break;

  case GNUNET_FS_STATUS_DOWNLOAD_ERROR:
    if (0 != isatty (1))
      fprintf (stdout, "\n");
    fprintf (stderr,
             _ ("Error downloading: %s.\n"),
             info->value.download.specifics.error.message);
    GNUNET_SCHEDULER_shutdown ();
    break;

  case GNUNET_FS_STATUS_DOWNLOAD_COMPLETED:
    s = GNUNET_STRINGS_byte_size_fancy (
      info->value.download.completed * 1000
      / (info->value.download.duration.rel_value_us + 1));
    if (0 != isatty (1))
      fprintf (stdout, "\n");
    fprintf (stdout,
             _ ("Downloading `%s' done (%s/s).\n"),
             info->value.download.filename,
             s);
    GNUNET_free (s);
    if (info->value.download.dc == dc)
      GNUNET_SCHEDULER_shutdown ();
    break;

  case GNUNET_FS_STATUS_DOWNLOAD_STOPPED:
    if (info->value.download.dc == dc)
      GNUNET_SCHEDULER_add_now (&cleanup_task, NULL);
    break;

  case GNUNET_FS_STATUS_DOWNLOAD_ACTIVE:
  case GNUNET_FS_STATUS_DOWNLOAD_INACTIVE:
    break;

  default:
    fprintf (stderr, _ ("Unexpected status: %d\n"), info->status);
    break;
  }
  return NULL;
}


/**
 * Main function that will be run by the scheduler.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param c configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *c)
{
  struct GNUNET_FS_Uri *uri;
  char *emsg;
  enum GNUNET_FS_DownloadOptions options;

  if (NULL == args[0])
  {
    fprintf (stderr, "%s", _ ("You need to specify a URI argument.\n"));
    return;
  }
  uri = GNUNET_FS_uri_parse (args[0], &emsg);
  if (NULL == uri)
  {
    fprintf (stderr, _ ("Failed to parse URI: %s\n"), emsg);
    GNUNET_free (emsg);
    ret = 1;
    return;
  }
  if ((! GNUNET_FS_uri_test_chk (uri)) && (! GNUNET_FS_uri_test_loc (uri)))
  {
    fprintf (stderr, "%s", _ ("Only CHK or LOC URIs supported.\n"));
    ret = 1;
    GNUNET_FS_uri_destroy (uri);
    return;
  }
  if (NULL == filename)
  {
    fprintf (stderr, "%s", _ ("Target filename must be specified.\n"));
    ret = 1;
    GNUNET_FS_uri_destroy (uri);
    return;
  }
  cfg = c;
  ctx = GNUNET_FS_start (cfg,
                         "gnunet-download",
                         &progress_cb,
                         NULL,
                         GNUNET_FS_FLAGS_NONE,
                         GNUNET_FS_OPTIONS_DOWNLOAD_PARALLELISM,
                         parallelism,
                         GNUNET_FS_OPTIONS_REQUEST_PARALLELISM,
                         request_parallelism,
                         GNUNET_FS_OPTIONS_END);
  if (NULL == ctx)
  {
    fprintf (stderr, _ ("Could not initialize `%s' subsystem.\n"), "FS");
    GNUNET_FS_uri_destroy (uri);
    ret = 1;
    return;
  }
  options = GNUNET_FS_DOWNLOAD_OPTION_NONE;
  if (do_recursive)
    options |= GNUNET_FS_DOWNLOAD_OPTION_RECURSIVE;
  if (local_only)
    options |= GNUNET_FS_DOWNLOAD_OPTION_LOOPBACK_ONLY;
  dc = GNUNET_FS_download_start (ctx,
                                 uri,
                                 NULL,
                                 filename,
                                 NULL,
                                 0,
                                 GNUNET_FS_uri_chk_get_file_size (uri),
                                 anonymity,
                                 options,
                                 NULL,
                                 NULL);
  GNUNET_FS_uri_destroy (uri);
  if (dc == NULL)
  {
    GNUNET_FS_stop (ctx);
    ctx = NULL;
    return;
  }
  GNUNET_SCHEDULER_add_shutdown (&shutdown_task, NULL);
}


/**
 * The main function to download GNUnet.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc, char *const *argv)
{
  struct GNUNET_GETOPT_CommandLineOption options[] =
  { GNUNET_GETOPT_option_uint ('a',
                               "anonymity",
                               "LEVEL",
                               gettext_noop (
                                 "set the desired LEVEL of receiver-anonymity"),
                               &anonymity),

    GNUNET_GETOPT_option_flag (
      'D',
      "delete-incomplete",
      gettext_noop ("delete incomplete downloads (when aborted with CTRL-C)"),
      &delete_incomplete),

    GNUNET_GETOPT_option_flag (
      'n',
      "no-network",
      gettext_noop ("only search the local peer (no P2P network search)"),
      &local_only),
    GNUNET_GETOPT_option_string ('o',
                                 "output",
                                 "FILENAME",
                                 gettext_noop ("write the file to FILENAME"),
                                 &filename),
    GNUNET_GETOPT_option_uint (
      'p',
      "parallelism",
      "DOWNLOADS",
      gettext_noop (
        "set the maximum number of parallel downloads that is allowed"),
      &parallelism),
    GNUNET_GETOPT_option_uint (
      'r',
      "request-parallelism",
      "REQUESTS",
      gettext_noop (
        "set the maximum number of parallel requests for blocks that is allowed"),
      &request_parallelism),
    GNUNET_GETOPT_option_flag ('R',
                               "recursive",
                               gettext_noop (
                                 "download a GNUnet directory recursively"),
                               &do_recursive),
    GNUNET_GETOPT_option_increment_uint (
      'V',
      "verbose",
      gettext_noop ("be verbose (print progress information)"),
      &verbose),
    GNUNET_GETOPT_OPTION_END };

  if (GNUNET_OK != GNUNET_STRINGS_get_utf8_args (argc, argv, &argc, &argv))
    return 2;

  ret =
    (GNUNET_OK ==
     GNUNET_PROGRAM_run (
       argc,
       argv,
       "gnunet-download [OPTIONS] URI",
       gettext_noop (
         "Download files from GNUnet using a GNUnet CHK or LOC URI (gnunet://fs/chk/...)"),
       options,
       &run,
       NULL))
    ? ret
    : 1;
  GNUNET_free ((void *) argv);
  return ret;
}


/* end of gnunet-download.c */
